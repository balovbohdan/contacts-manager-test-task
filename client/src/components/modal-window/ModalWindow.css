.main {
    z-index: 99999;
    position: absolute;

    width: 100vw;
    height: 100vh;

    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    display: flex;
    align-items: center;
    justify-content: center;
}

.body {
    max-width: 450px;
    width: calc(100vw - 20px);

    min-height: 200px;
    max-height: calc(100vh - 40px);

    padding: 10px;
    margin: 20px auto;

    position: relative;
    border-radius: var(--border-radius);
    background-color: var(--color-light-blue);
    box-shadow: 0 20px 35px rgba(0, 0, 0, .3), 0 15px 12px rgba(0, 0, 0, .2);

    color: var(--color-white);
}

.head {
    position: relative;
    padding-bottom: 10px;

    display: flex;
    align-items: center;
    justify-content: flex-start;
}

.closeBtn {
    position: absolute;

    top: -20px;
    right: -20px;

    width: 40px;
    height: 40px;

    display: flex;
    align-items: center;
    justify-content: center;

    cursor: pointer;
    border-radius: 50%;
    transition: var(--trans);
    background-color: var(--color-red);
    box-shadow: 0 3px 10px rgba(0, 0, 0, .3), 0 5px 10px rgba(0, 0, 0,.2);
}

.closeBtn:hover {
    transform: rotate(10deg) scale(.9);
}

.closeBtn_img {
    width: calc(100% - 20px);
    height: calc(100% - 20px);

    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
    background-image: url(~@img/close.svg);
}