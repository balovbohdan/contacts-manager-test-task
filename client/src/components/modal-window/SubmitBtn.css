.submitBtn {
    cursor: pointer;
    padding: 10px 15px;
    transition: var(--trans);
    border-radius: var(--border-radius);
    background-color: rgba(0, 0, 0, .2);
}

.submitBtn:hover {
    background-color: rgba(0, 0, 0, .3);
}

.submitBtn__disabled {
    cursor: not-allowed;
}

.submitBtn__disabled:hover {
    background-color: rgba(0, 0, 0, .2);
}