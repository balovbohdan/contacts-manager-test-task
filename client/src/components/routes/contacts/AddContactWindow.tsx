import * as React from 'react';

import {T as Actions} from '@storehouse/actions/contacts';
import {Contacts} from '@lib/entities/contacts/contact/types';
import {AddContactWindow as Base} from '@components/contacts-manager';

type Props = {
    active:boolean;
    contacts:Contacts;

    close:()=>void;
    addContact:Actions.AddContact;
};

export const AddContactWindow = (props:Props) =>
    !props.active
        ? null
        : <Base
            close={props.close}
            contacts={props.contacts}
            addContact={props.addContact}/>;