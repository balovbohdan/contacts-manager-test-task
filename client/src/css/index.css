@import url(reset.css);
@import url(fonts.css);
@import url(titles.css);
@import url(constants.css);
@import url(main.css);