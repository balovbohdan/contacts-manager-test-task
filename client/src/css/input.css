.label {
    font-size: .9rem;
}

.input {
    width: 100%;
    padding: 5px 0;
    border-bottom: 2px solid var(--color-yellow);

    font-size: 1.5rem;
    color: var(--color-white);
}

@media (max-width: 600px) {
    .input {
        font-size: 1.2rem;
    }

    .label {
        font-size: .8rem;
    }
}

@media (max-width: 400px) {
    .input {
        font-size: 1.1rem;
    }

    .label {
        font-size: .7rem;
    }
}