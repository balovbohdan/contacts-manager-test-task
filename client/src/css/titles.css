h1, h2, h3, h4, h5, h6 {
    font-size: 100%;
    font-family: var(--font-main);
}

h1 {
    font-size: 40px;
}

h2 {
    font-size: 35px;
}

h3 {
    font-size: 30px;
}

h4 {
    font-size: 25px;
}

h5 {
    font-size: 20px;
}

h6 {
    font-size: 17px;
}

@media (max-width: 550px) {
    h1 {
        font-size: 35px;
    }

    h3 {
        font-size: 25px;
    }
}

@media (max-width: 400px) {
    h1 {
        font-size: 30px;
    }

    h3 {
        font-size: 20px;
    }
}