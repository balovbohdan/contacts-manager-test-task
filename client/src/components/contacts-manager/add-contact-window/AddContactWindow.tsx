import * as React from 'react';

import {genRndNumber} from '@lib/math';
import {Contact} from '@lib/entities/contacts/contact/types';

import {ModalWindow} from '@components/modal-window';
import {T as Actions} from '@storehouse/actions/contacts';

import {T, Form} from '../form';

type Props = {
    close:Close;
    addContact:Actions.AddContact;
};

type Close = ()=>void;

export const AddContactWindow = (props:Props) => {
    const submit = createSubmit(props.close, props.addContact);

    return (
        <ModalWindow close={props.close} title='Add contact'>
            <Form submit={submit}/>
        </ModalWindow>
    );
};

const createSubmit = (close:Close, addContact:Actions.AddContact) =>
    (formData:T.FormData) => {
        const contact = createContactFromFormData(formData);

        addContact({ contact });
        close();
    };

const createContactFromFormData = (formData:T.FormData):Contact => {
    const id = genRndNumber();

    return Object.assign({}, formData, { id });
};