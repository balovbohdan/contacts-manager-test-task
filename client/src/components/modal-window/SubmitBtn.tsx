import * as React from 'react';
import classnames from 'classnames';

import * as css from './SubmitBtn.css';

type Props = {
    submit:()=>void;
    isDataValid:boolean;
};

export const SubmitBtn = ({submit, isDataValid}:Props) => {
    const doSubmit = () => isDataValid && submit();

    const className = classnames({
        [css.submitBtn]: true,
        [css.submitBtn__disabled]: !isDataValid
    });

    return <div onClick={doSubmit} className={className}>Submit</div>;
};