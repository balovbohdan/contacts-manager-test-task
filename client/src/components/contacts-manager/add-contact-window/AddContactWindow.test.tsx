import * as React from 'react';
import {mount, shallow} from 'enzyme';

import {AddContactWindow} from './AddContactWindow';
import * as windowCss from '../../modal-window/ModalWindow.css';
import * as submitBtnCss from '../../modal-window/SubmitBtn.css';

describe('AddContactWindow', () => {
    const propsDef = {
        close: jest.fn(),
        addContact: jest.fn(),
        contacts: [{
            id: 1,
            name: 'name',
            img: '/user.svg',
            phone: '+380951419184'
        }]
    };

    it('Should match snapshot', () => {
        const props = { ...propsDef };
        const window = shallow(<AddContactWindow {...props}/>);

        expect(window).toMatchSnapshot();
    });

    it('Should react on close btn click', () => {
        const props = {
            ...propsDef,
            close: jest.fn()
        };

        const window = mount(<AddContactWindow {...props}/>);

        window.find(`.${windowCss.closeBtn}`).simulate('click');
        expect(props.close).toBeCalledTimes(1);
    });

    it('Should close after empty space click', () => {
        const props = {
            ...propsDef,
            close: jest.fn()
        };

        const window = mount(<AddContactWindow {...props}/>);

        window.find(`.${windowCss.main}`).at(0).simulate('click');
        expect(props.close).toBeCalledTimes(1);
    });

    it('Should react on submit btn click', () => {
        const props = {
            ...propsDef,
            close: jest.fn(),
            addContact: jest.fn(),
            def: {
                name: 'name',
                phone: '+380951410191'
            }
        };

        const window = mount(<AddContactWindow {...props}/>);

        window.find(`.${submitBtnCss.submitBtn}`).simulate('click');

        expect(props.close).toBeCalledTimes(1);
        expect(props.addContact).toBeCalledTimes(1);
    });

    it('Should not react on submit btn click when data is invalid', () => {
        const props = {
            ...propsDef,
            close: jest.fn(),
            addContact: jest.fn(),
            def: {
                name: '',
                phone: '+380913231065'
            }
        };

        const window = mount(<AddContactWindow {...props}/>);

        window.find(`.${submitBtnCss.submitBtn}`).simulate('click');

        expect(props.close).toBeCalledTimes(0);
        expect(props.addContact).toBeCalledTimes(0);
    });
});