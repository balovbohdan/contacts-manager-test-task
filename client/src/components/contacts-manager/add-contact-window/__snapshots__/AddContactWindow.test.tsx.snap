// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddContactWindow Should match snapshot 1`] = `
<Component
  close={[MockFunction]}
  title="Add contact"
>
  <Component
    submit={[Function]}
  />
</Component>
`;
